### PR TITLE
Provider

### PR DESCRIPTION
**Pull request by alentner**
Drafted a new feature to implement specific providers for both Namespace and Configuration.

The feature also provides method for trimming duplicates while maintaining validity of data structure in the builder provider.

### CODE REVIEW

**Comment by alentner** [comment]
This draft includes the changes in the Duplicates pull request.

**Comment by glentner** [comment]
I like where we're going with this. I think a `provider` sub-package with varying use-cases that implement more complicated features by extending the `Namespace` and/or `Configuration` is the right way to go.

Do we want to have a top-level `provider` sub-package or do we want to have `cmdkit.config.provider.builder` instead?

**Comment by glentner** [comment]
The unit tests for `BuilderConfiguration` do not all pass. I think the issue might be the `pop` used, which are not implemented for the base `Configuration`. Does `BuilderConfiguration` need to implement these?

**Review by alentner — COMMENTED** [review]
Ready for merging with develop

File: cmdkit/config.py [hunk 1]
```diff
@@ -629,7 +629,8 @@ def duplicates(self) -> Dict[str, Dict[str, List[Tuple[str, ...]]]]:
             >>> cfg.duplicates()
             {'x': {'one': [('a',), ('b',)], 'two': [('b',)]}, 'z': {'one': [('b',)], 'two': [('b',)]}}
         """
-        tips = [tip for _, (*_, tip) in _find_the_leaves(self.namespaces)]
+        namespaces = Namespace({**self.namespaces, '_': self.local})
+        tips = [tip for _, (*_, tip) in _find_the_leaves(namespaces)]
```
Comment (glentner) [inline cmdkit/config.py:633]: Fixes bug with local

**Review by glentner — APPROVED** [review]
Fixes bug with original implementation. The use of `del` is unfortunate, but better than the alternative -- it's an old code but it checks out.

**Pull request merged**